app/registries/foreman/plugin/medium_providers_registry.rb

```diff
@@ -18,24 +19,28 @@ def register(provider)
         provider
       end
 
+      def unregister(provider)
+        @providers.delete(provider.to_s)
+      end
+
       # Find a provider that can provide a URI for a given entity.
       def find_provider(entity)
         raise Foreman::Exception.new('Must supply an entity to find a medium provider') unless entity
 
         provider_instances = providers.map { |provider| provider.constantize.new(entity) }
         valid_providers = provider_instances.select { |provider| provider.valid? }
         if valid_providers.count > 1
-          raise Foreman::Exception.new(
-            'Found more than one provider for %{entity}. Valid providers: %{providers}',
+          logger.error(
+            'Found more than one provider for %{entity}. Valid providers: %{providers}' %
             {
               entity: entity,
               providers: providers.map { |provider| provider.class.friendly_name }
             })
         end
 
         unless valid_providers.present?
```

Comment:
You could return early here if you want.

`return valid_providers.first if valid_providers.present?`

Reply:
`valid_providers.first` will return nil if nothing's found, so I honestly prefer the obvious return at the end, unless you'd really prefer it that way